README.md
# postcss-hairline

[![NPM version][npm-image]][npm-url] [![Build Status][ci-status-image]][ci-status-url] [![Coverage Status][coverage-status-image]][coverage-status-url] [![Known Vulnerabilities][vulnerabilities-status-image]][vulnerabilities-status-url] [![changelog][changelog-image]][changelog-url] [![license][license-image]][license-url]

[vulnerabilities-status-image]: https://snyk.io/test/npm/postcss-hairline/badge.svg
[vulnerabilities-status-url]: https://snyk.io/test/npm/postcss-hairline
[ci-status-image]: https://travis-ci.org/ufologist/postcss-hairline.svg?branch=master
[ci-status-url]: https://travis-ci.org/ufologist/postcss-hairline
[coverage-status-image]: https://coveralls.io/repos/github/ufologist/postcss-hairline/badge.svg?branch=master
[coverage-status-url]: https://coveralls.io/github/ufologist/postcss-hairline
[npm-image]: https://img.shields.io/npm/v/postcss-hairline.svg?style=flat-square
[npm-url]: https://npmjs.org/package/postcss-hairline
[license-image]: https://img.shields.io/github/license/ufologist/postcss-hairline.svg
[license-url]: https://github.com/ufologist/postcss-hairline/blob/master/LICENSE
[changelog-image]: https://img.shields.io/badge/CHANGE-LOG-blue.svg?style=flat-square
[changelog-url]: https://github.com/ufologist/postcss-hairline/blob/master/CHANGELOG.md

[![npm-image](https://nodei.co/npm/postcss-hairline.png?downloads=true&downloadRank=true&stars=true)](https://npmjs.com/package/postcss-hairline)

PostCSS plugin for transform your border to retina [hairline](https://github.com/ufologist/hairline).

## Example

before:

```css
.foo {
    color: white;
    border: 1px solid red; /* hairline */
    border-radius: 20px; /* hairline */
}
```

after:

```css
.foo {
    color: white;
    border-radius: 10px; /* no */
    position: relative; 
}

.foo::after {
    border: 1px solid red; /* no */
    border-radius: 20px; /* no */
}

.foo::after {
    content: '';
    position: absolute;
    top: -50%;
    bottom: -50%;
    left: -50%;
    right: -50%;
    -webkit-transform: scale(0.5);
            transform: scale(0.5);
    pointer-events: none;
}
```

## Usage

1. install plugin

   ```
   npm install --save-dev postcss-hairline
   ```

2. code `border` or `border-radius` style as usual

   ```css
   .foo {
       color: white;
       border: 1px solid red;
       border-radius: 20px;
   }
   ```

3. add `/* hairline */` comment

   ```css
   .foo {
       color: white;
       border: 1px solid red; /* hairline */
       border-radius: 20px; /* hairline */
   }
   ```

   support comment
   * `/* hairline */` default create `::after` pseudo style
   * `/* hairline-before */` specify create `::before` pseudo style
   * `/* hairline-after */` specify create `::after` pseudo style

4. use plugin

   ```
   postcss([
       require('postcss-hairline')
   ]);
   ```

   [PostCSS plugin usage](https://github.com/postcss/autoprefixer#usage)

## Options

```javascript
require('postcss-hairline')({
    pxComment: 'no'
})
```

Available options are:
* `pxComment` (string): border px unit comment. Default: `no`.

## Reference

* [postcss-retina](https://github.com/Ziphwy/postcss-retina)

CHANGELOG.md
# CHANGELOG

* v1.0.0 2019-5-15

  * 初始版本

test/index.test.js
const postcss = require('postcss');
const fs = require('fs');
const path = require('path');

const plugin = require('../src/index.js');

const readFile = (filePath, ...args) => new Promise((resolve, reject) => {
    fs.readFile(path.resolve(__dirname, filePath), ...args, (err, data) => {
        if (err) {
            reject();
        } else {
            resolve(data);
        }
    });
});

async function run(input, opts) {
    return postcss([plugin(opts)]).process(input).then((result) => {
               expect(result.css).toMatchSnapshot();
               expect(result.warnings().length).toBe(0);
           });
}

it('transform border with border-radius', async () => {
    const input = await readFile('./case/border-width-radius.css');
    run(input);
});

src/index.js
const postcss = require('postcss');
const {
    getPseudoSelector,
    unique
} = require('./helper');

const hairlineReg = /^hairline(?:-?)(\S*)/;
const borderReg = /border(?:-top|-bottom|-left|-right|-image)?(?:-width)?$/;
const borderRadiusReg = /border-radius/;
const unitPxReg = /([0-9\.]+)(px)/;
const unitRemReg = /([0-9\.]+)(rem)/;
const pseudoReg = /:(?:before|after)$/;

const hairline = postcss.plugin('postcss-hairline', (options = {}) => (root) => {
    const {
        pxComment = 'no' // 添加注释避免边框宽度被类似 postcss-px2rem 这样的插件转成 rem 单位
    } = options;

    const hairlineRuleList = [];
    const selectors = [];

    root.walkRules((rule) => {
        const borderDecls = [];
        let borderPseudo = 'after';
        let isPositionStatic = true;

        rule.walkDecls((decl) => {
            if (borderReg.test(decl.prop) && unitPxReg.test(decl.value)) {
                const next = decl.next();
                if (next && next.type === 'comment' && hairlineReg.test(next.text.trim())) {
                    var pseudo = next.text.trim().match(hairlineReg)[1];
                    if (pseudo) {
                        borderPseudo = pseudo;
                    }

                    decl.remove();
                    borderDecls.push(decl);
                }
            }

            if (borderRadiusReg.test(decl.prop)) {
                let borderRadiusDecl;

                const next = decl.next();
                if (next && next.type === 'comment' && hairlineReg.test(next.text.trim())) {
                    // 容器圆角为R, 伪元素圆角为2R
                    if (unitPxReg.test(decl.value)) {
                        // px 单位时, 例如 750 设计稿, 20px 的圆角
                        // 容器圆角应为10px(书写20px), 伪元素圆角应为20px
                        borderRadiusDecl = decl;

                        decl.remove();
                        // 调整容器的圆角
                        rule.append(decl.clone({
                            value: decl.value.replace(unitPxReg, ($0, $1, $2) => `${parseFloat($1) / 2}${$2}`)
                        }));
                        rule.append({
                            text: pxComment
                        });
                    } else if (unitRemReg.test(decl.value)) {
                        // rem 单位时, 例如 750 设计稿, 20px 的圆角
                        // 容器圆角应为2rem, 伪元素圆角应为4rem
                        borderRadiusDecl = decl.clone({
                            value: decl.value.replace(unitRemReg, ($0, $1, $2) => `${parseFloat($1) * 2}${$2}`)
                        });
                    }

                    if (borderRadiusDecl) {
                        borderDecls.push(borderRadiusDecl);
                    }
                }
            }

            if (decl.prop === 'position' && decl.value !== 'static') {
                isPositionStatic = false;
            }
        });

        const slts = rule.selectors.filter(slt => !pseudoReg.test(slt));
        if (borderDecls.length <= 0  || slts.length <= 0) {
            return;
        }

        // 边框个性化样式
        const hairlineRule = postcss.rule({
            selector: getPseudoSelector(slts, borderPseudo, true)
        });
        borderDecls.forEach((decl) => {
            hairlineRule.append(decl);
            hairlineRule.append({
                text: pxComment
            });
        });
        hairlineRuleList.push(hairlineRule);

        // 调整当前的规则
        if (isPositionStatic) {
            rule.append({
                prop: 'position',
                value: 'relative'
            });
        }

        [].push.apply(selectors, getPseudoSelector(slts, borderPseudo));
    });

    hairlineRuleList.forEach((rule) => {
        root.append(rule);
    });

    // 创建伪元素的样式的统一样式
    if (selectors.length > 0) {
        const slts = unique(selectors);
        root.append(`
            ${slts.join(',')} {
                content: '';
                position: absolute;
                top: -50%;
                bottom: -50%;
                left: -50%;
                right: -50%;
                -webkit-transform: scale(0.5);
                        transform: scale(0.5);
                pointer-events: none;
            }
        `);
    }
});

module.exports = hairline;
